File: 2021/11/02/main.py
class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]
        queue, paths, walked = [], [], 0

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    walked |= 1 << (i*n+j)
                    queue.append([i, j])
                    paths.append(1<<(i*n+j))
                elif grid[i][j] == 0:
                    walked |= 1 << (i*n+j)
                
        count = 0

        while len(queue) > 0:
            size = len(queue)
            for s in range(size):
                x, y = queue[s][0], queue[s][1]
                path = paths[s]
                for k in range(4):
                    i, j = x+dx[k], y+dy[k]
                    if i < 0 or m <= i or j < 0 or n <= j or path&(1<<(i*n+j)) != 0:
                        continue
                    
                    if grid[i][j] == 0:
                        queue.append([i, j])
                        paths.append(path|(1<<(i*n+j)))
                    elif grid[i][j] == 2:
                        if path == walked:
                            count += 1
            queue = queue[size:]
            paths = paths[size:]

        return count

File: 2021/11/02/main.go
package main

var dx = []int{0, 0, 1, -1}
var dy = []int{1, -1, 0, 0}

func uniquePathsIII(grid [][]int) int {

	m, n := len(grid), len(grid[0])

	queue := make([][2]int, 0, m*n*10)
	paths := make([]int, 0, m*n*10)
	walked := 0

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			switch grid[i][j] {
			case 0:
				walked |= 1 << uint(i*n+j)
			case 1:
				walked |= 1 << uint(i*n+j)
				queue = append(queue, [2]int{i, j})
				paths = append(paths, 1<<uint(i*n+j))
			}
		}
	}

	count := 0

	for len(queue) > 0 {
		size := len(queue)
		for s := 0; s < size; s++ {
			x, y := queue[s][0], queue[s][1]
			path := paths[s]
			for k := 0; k < 4; k++ {
				i, j := x+dx[k], y+dy[k]
				if i < 0 || m <= i ||
					j < 0 || n <= j ||
					path&(1<<uint(i*n+j)) != 0 {
					continue
				}
				switch grid[i][j] {
				case 0:
					queue = append(queue, [2]int{i, j})
					paths = append(paths, path|(1<<uint(i*n+j)))
				case 2:
					if path == walked {
						count++
					}
				}
			}
		}
		queue = queue[size:]
		paths = paths[size:]
	}

	return count
}
